core/templates/filters/translate.pipe.spec.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the translate pipe.
+ */
+
+import { async, TestBed } from '@angular/core/testing';
+import { EventEmitter, ChangeDetectorRef } from '@angular/core';
+
+import { TranslateService } from 'services/translate.service';
+import { TranslatePipe } from './translate.pipe';
+import { UtilsService } from 'services/utils.service';
+
+
+class MockChangeDetectorRef {
+  markForCheck() {
+    return true;
+  }
+}
+
+class MockTranslateService {
+    langChange = new EventEmitter();
+    templateMatcher: RegExp = /\<\[\s?([^{}\s]*)\s?\]\>/g;
+    get onLangChange() {
+      return this.langChange;
+    }
+    translations = {
+      I18n_t_1: 'Hello',
+      I18n_t_2: 'Hello <[val]>'
+    };
+    private interpolateString(expr: string, params?: Object | undefined) {
+      if (!params) {
+        return expr;
+      }
+      return expr.replace(this.templateMatcher,
+        (substring: string, b: string) => {
+          return params[b];
+        });
+    }
+
+    getInterpolatedString(key: string,
+        interpolateParams?: Object) {
+      const str = this.translations[key];
+      if (!interpolateParams) {
+        return str;
+      }
+
+      return this.interpolateString(str, interpolateParams);
+    }
+}
+
+describe('TranslatePipe', () => {
+  let pipe: TranslatePipe;
+  let translate: TranslateService;
+  let changeDecRef: ChangeDetectorRef;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        TranslatePipe
+      ],
+      providers: [
+        { provide: TranslateService, useClass: MockTranslateService },
+        { provide: ChangeDetectorRef, useClass: MockChangeDetectorRef}
+      ]
+    }).compileComponents();
+    translate = TestBed.get(TranslateService);
+    changeDecRef = TestBed.get(ChangeDetectorRef);
+    pipe = new TranslatePipe(translate, changeDecRef, new UtilsService());
+  }));
+
+  it('should translate', () => {
+    expect(pipe.transform('I18n_t_1')).toBe('Hello');
+    expect(pipe.transform('I18n_t_2', {val: 'World'})).toBe('Hello World');
```

Comment:
What are we doing to sanitize/escape the values in params?

Add tests such that we try to inject script tags etc. into params via a rogue translation, see what happens.

Reply:
As discussed before(https://github.com/oppia/oppia/pull/9272#discussion_r429508417) we don't sanitize anything here. Angular does the sanitization. I believe internally they have a very narrow list of safe "values". By default, everything is treated as unsafe by angular. In case the need arises for us to have some custom-tags in our translations we will have to write the code to mark them as safe. But for now, we don't have any such translations.

Reply:
![image](https://user-images.githubusercontent.com/23002114/83799628-bb558900-a6c3-11ea-8a7e-00229db4cf0b.png)

Added `<script>alert('Hello');</script>`

Angular removed that
![image](https://user-images.githubusercontent.com/23002114/83799811-0079bb00-a6c4-11ea-88b8-20415cd470ce.png)

Reply:
Cool, that's great. Please make sure cases like these are covered in the unit tests.

Reply:
Yeah, I am writing a test at the moment. 😁

Reply:
https://github.com/oppia/oppia/pull/9398/files#r435528580